sdk/iot/core/tests/cmocka/az_iot_telemetry_tests.c

```diff
@@ -0,0 +1,283 @@
+#include <az_iot_hub_client.h>
+#include <az_span.h>
+
+#include <stdio.h>
+
+#include <setjmp.h>
+#include <stdarg.h>
+#include <stddef.h>
+#include <stdint.h>
+
+#include <cmocka.h>
+
+#define TEST_CORRECT_TOPIC_PREFIX "devices/"
+#define TEST_CORRECT_TOPIC_SUFFIX "/messages/events/"
+#define TEST_CORRECT_TOPIC_MODULES_MID "/modules/"
+
+#define TEST_FQDN "myiothub.azure-devices.net"
+#define TEST_DEVICE_ID "my_device"
+#define TEST_MODULE_ID "my_module_id"
+#define TEST_USER_AGENT "api-version=2017-11-08-preview&DeviceClientType=iothubclient"
+#define TEST_PARAMS "key=value&key_two=value2"
+#define TEST_MQTT_SPAN_BUFFER_SIZE 100
+
+#define TEST_BAD_SPAN \
+  { \
+    ._internal = { .ptr = (uint8_t*)0x7FFFFFFF, .length = -1, .capacity = 10  }, \
+  }
+#define TEST_EMPTY_OPTIONS \
+  { \
+    .module_id = AZ_SPAN_LITERAL_NULL, .user_agent = AZ_SPAN_LITERAL_NULL \
+  }
+#define TEST_VALID_OPTIONS_BOTH \
+  { \
+    .module_id = AZ_SPAN_LITERAL_FROM_STR(TEST_MODULE_ID), \
+    .user_agent = AZ_SPAN_LITERAL_FROM_STR(TEST_USER_AGENT) \
+  }
+#define TEST_VALID_OPTIONS_MODULE_ID \
+  { \
+    .module_id = AZ_SPAN_LITERAL_FROM_STR(TEST_MODULE_ID), .user_agent = AZ_SPAN_LITERAL_NULL \
+  }
+#define TEST_VALID_OPTIONS_USER_AGENT \
+  { \
+    .module_id = AZ_SPAN_LITERAL_NULL, .user_agent = AZ_SPAN_LITERAL_FROM_STR(TEST_USER_AGENT) \
+  }
+
+static const az_iot_hub_client_properties g_test_params
+    = { ._internal
+        = { .properties = AZ_SPAN_LITERAL_FROM_STR(TEST_PARAMS), .current_property = 0 } };
+
+static const char g_test_correct_topic_no_options_no_params[]
```

Comment:
These are extremely hard to read for a unit test. Instead you should define the actual strings.

Reply:
I think I have my preferences here. I can change them if you feel strongly but this allows me to compose them easily and make sure copy paste errors don't happen.